### ISSUE TITLE
Setting up WebSocket protocol

### ISSUE THREAD

**Issue by snezhana-krass** [lightbody/browsermob-proxy#408]
Hi,
when I use browsermob-proxy all connections goes through https:// even WebSocket methods. 
For example: GET /echo HTTP/1.1 returns WebSocket-Location: wss://site.com/echo/123/websocket, after that initiates GET **wss**://site.com/echo/123/websocket with response code 101 switching protocols. But with browsermob-proxy initiates request GET **https**://site.com/echo/123/websocket

I've try to change protocol with addRequestFilter() but there is only http version of protocol:

```
proxy.addRequestFilter(new RequestFilter() {
   @Override
    public HttpResponse filterRequest(HttpRequest request, HttpMessageContents contents,    HttpMessageInfo messageInfo) {
	   if (request.getUri().contains("echo/") && request.getUri().contains("/websocket")) {
		   request.setProtocolVersion(new HttpVersion("WSS/1.1", true));
	   }
	   return null;
   }
});
```

There is no info about WebSocket and browsermob-proxy just this self-reply https://groups.google.com/forum/#!topic/browsermob-proxy/JzTHxje0qP4

I need proxy to monitor HTTP and WebSocket, there are a lot of autotests that I could make with it, but if I run tests with WebSocket it does not work at all.

Kind regards!

**Comment by jekh** [comment]
BMP doesn't currently support WebSockets because LittleProxy doesn't support WebSockets (AFAIK). I am not familiar enough with WebSockets to implement them myself, but I would welcome the efforts of anyone who would like to add them to LittleProxy!

**Comment by scythargon** [comment]
Hi! May be there is a way do disable interception for websocket connections at all?
BTW I'm using python bindings https://browsermob-proxy-py.readthedocs.io/en/stable/

**Comment by snezhana-krass** [comment]
Hi, I use PAC file to 
// Start BrowserMobProxy server
BrowserMobProxy proxyServer = new BrowserMobProxyServer();
proxyServer.setTrustAllServers(true);
proxyServer.setRequestTimeout(120, TimeUnit.SECONDS);
proxyServer.setConnectTimeout(120, TimeUnit.SECONDS);
proxyServer.enableHarCaptureTypes(CaptureType.getAllContentCaptureTypes());
 proxyServer.start(0);

// Get BrowserMobProxy server port for PAC file
int proxyPort = proxyServer.getPort();

// Create PAC file to send WebSocket requests direct but other protocols throught BrowserMobProxy server
String pacFunction = "function FindProxyForURL(url, host) { if (url.substring(0, 3) === \"ws:\" || url.substring(0, 4) === \"wss:\") { " +
                "return \"DIRECT\"; } else { return \"PROXY localhost:" + proxyPort + "\"; } }";
String pacLocation = String.format("target/pac/websocket-proxy.pac", proxyPort); 
writeStringToFile(new File(pacLocation), pacFunction, "UTF-8");

// Set PAC file path to Selenium proxy
Proxy seleniumProxy = new Proxy();
seleniumProxy.setProxyType(Proxy.ProxyType.PAC);
seleniumProxy.setProxyAutoconfigUrl(String.format(System.getProperty("base-dir"), pacLocation));

// Add Selenium proxy to DesiredCapabilities
DesiredCapabilities desiredCapabilities = capabilities;
desiredCapabilities.setCapability(CapabilityType.PROXY, seleniumProxy);
desiredCapabilities.setCapability(CapabilityType.ACCEPT_SSL_CERTS, true);

// Start driver with desiredCapabilities
WebDriver driver = new ChromeDriver(capabilities);

But it work only in Chrome and does not work in Firefox.

**Comment by scythargon** [comment]
Hi, Snezhana!
Yup, looks like something I might use, thanks! I'm trying to figure out how to do something similar with this python wrapper.
Привет соотечественникам:)

**Comment by scythargon** [comment]
BTW where did you find the information about PAC files, how should your JS function look like, and what it should return?

**Comment by snezhana-krass** [comment]
I've found out this answer
https://groups.google.com/forum/m/#!topic/browsermob-proxy/JzTHxje0qP4


ср, 15 февр. 2017 г. в 19:35, scythargon <notifications@github.com>:

**Comment by affinage** [comment]
@scythargon did you find the way it works on python?

**Comment by scythargon** [comment]
@affinage unfortunately, nope

**Comment by Ravichandran-3112** [comment]
Hi snezhana-krass
What is base-dir here, i am getting null for base-dir
seleniumProxy.setProxyAutoconfigUrl(String.format(System.getProperty("base-dir"), pacLocation));

**Comment by twistedpair** [comment]
Note: while the PAC file appraoch this will work for `ws://` and `wss://` requests, it will not work for protocol upgrades over HTTP, e.g. like the Intercom does with the header `Upgrade: websocket`. Those connections will fail.

**Comment by snezhana-dorogova** [comment]
@Ravichandran-3112 String.format(System.getProperty("base-dir"), pacLocation) - this is equal to the path for PAC file location. I use "base-dir" in Maven pom.xml file for project files paths.

**Comment by ineverov** [comment]
@jekh Looks like littleproxy supports WebSockets already https://github.com/adamfisk/LittleProxy/releases/tag/littleproxy-1.1.0
Could somebody take a look at this issue?

**Comment by alex4200** [comment]
@snezhana-krass I tried to use your answer, and so I created a `poc` file with the following content

```
function FindProxyForURL(url, host) 
{ 
    if (url.substring(0, 3) === "ws:" || url.substring(0, 4) === "wss:") 
    { 
        return "DIRECT"; 
    } else { 
        return "PROXY 0.0.0.0:8080"; 
    }   
}
```

and a simple python script (for firefox, however):

```
proxy = Client("localhost:8080", {})
proxy.new_har("test")

profile = webdriver.FirefoxProfile()
selenium_proxy = webdriver.Proxy()
selenium_proxy.proxy_type = webdriver.common.proxy.ProxyType.PAC
selenium_proxy.proxyAutoconfigUrl = "/path/to/poc-file/proxy.poc"
profile.set_proxy(selenium_proxy) 
driver = webdriver.Firefox(firefox_profile=profile)

driver.get("http://www.python.org")
assert "Python" in driver.title
elem = driver.find_element_by_name("q")
elem.clear()
elem.send_keys("pycon")
elem.send_keys(Keys.RETURN)
assert "No results found." not in driver.page_source
driver.close()
print (proxy.har)
```

The example 'works', but the har file has no entries at all. 

When I use the proxy directly from the `browsermob-proxy` object (`selenium_proxy = proxy.selenium_proxy()`) I get a HAR list with many entries. 

But, like you, I want to solve the issue with the websockets I have with _jupyter notebooks_...

I also tried to use this approach for chrome, but no chromedriver seems to work for the basic approach with the simple proxy. So I am blocked with chrome as well (bug report filed). 

Any suggestions?

**Comment by alex4200** [comment]
I think I found a workaround. The idea is to define a PAC file to avoid any websocket requests through browsermob-proxy. My PAC file looks like follows

```
// Calomel.org Proxy Auto-Config
//
 
//
// Define the network paths (direct, proxy and deny)
//
 
// Default connection
var direct = "DIRECT";
 
// Alternate Proxy Server
var proxy = "PROXY 0.0.0.0:SHANGHAI";
 
// Default localhost for denied connections
var deny = "PROXY 0.0.0.0:65535";
 
//
// Proxy Logic
//
 
function FindProxyForURL(url, host)
{
    if (url.substring(0, 3) === "ws:" || url.substring(0, 4) === "wss:")
    {
        return direct;
    } else {
        return proxy;
    }
 
    return direct;
}
```

where I replace `SHANGHAI` with the port number of the current proxy (`client.port` where `client=bmp_server.create_proxy()`). Then in the python code I configure the webdriver as follows:

```
from selenium import webdriver
import pathlib # Must be installed via 'pip install pathlib' for python<3.4
 
selenium_proxy = webdriver.Proxy()       
selenium_proxy.proxy_type = webdriver.common.proxy.ProxyType.PAC
selenium_proxy.proxy_autoconfig_url = pathlib.Path(path_to_pac_file).as_uri()
 
profile = webdriver.FirefoxProfile()
profile.set_proxy(selenium_proxy)
 
driver = webdriver.Firefox(firefox_profile=profile)
```

where `path_to_pac_file` is the path to the PAC file.

**Comment by yippeykeiyay** [comment]
Hi,

I am having this problem currently with BMP v2.1.4. I'm observing the initial websocket connection attempt over `http://` and `https://` but nothing after.

I'm trying the PAC approach, but believe that this problem should have been fixed in BMP v2.1.0-beta-6?

Has anyone ran into this recently and found a solution? Any help would be appreciated!

Thanks!

**Comment by apreg** [comment]
> Hi, I use PAC file
> // Start BrowserMobProxy server
> BrowserMobProxy proxyServer = new BrowserMobProxyServer();
> proxyServer.setTrustAllServers(true);
> proxyServer.setRequestTimeout(120, TimeUnit.SECONDS);
> proxyServer.setConnectTimeout(120, TimeUnit.SECONDS);
> proxyServer.enableHarCaptureTypes(CaptureType.getAllContentCaptureTypes());
> proxyServer.start(0);
> 
> // Get BrowserMobProxy server port for PAC file
> int proxyPort = proxyServer.getPort();
> 
> // Create PAC file to send WebSocket requests direct but other protocols throught BrowserMobProxy server
> String pacFunction = "function FindProxyForURL(url, host) { if (url.substring(0, 3) === "ws:" || url.substring(0, 4) === "wss:") { " +
> "return "DIRECT"; } else { return "PROXY localhost:" + proxyPort + ""; } }";
> String pacLocation = String.format("target/pac/websocket-proxy-%s.pac", proxyPort);
> writeStringToFile(new File(pacLocation), pacFunction, "UTF-8");
> 
> // Set PAC file path to Selenium proxy
> Proxy seleniumProxy = new Proxy();
> seleniumProxy.setProxyType(Proxy.ProxyType.PAC);
> seleniumProxy.setProxyAutoconfigUrl(String.format(System.getProperty("base-dir"), pacLocation));
> 
> // Add Selenium proxy to DesiredCapabilities
> DesiredCapabilities desiredCapabilities = capabilities;
> desiredCapabilities.setCapability(CapabilityType.PROXY, seleniumProxy);
> desiredCapabilities.setCapability(CapabilityType.ACCEPT_SSL_CERTS, true);
> 
> // Start driver with desiredCapabilities
> WebDriver driver = new ChromeDriver(capabilities);
> 
> But it works only in Chrome and does not work in Firefox.

Thank you for this workaround. I had to tweak it a little bit since Chrome and Windows blocks the file:// schema usage for pac file url and I did not want to start a webserver to host the pac file. I found the solution here: https://superuser.com/questions/1405499/chrome-72-seems-to-no-longer-accept-proxy-pac-url
`--proxy-pac-url='data:application/x-javascript-config;base64,'$(base64 -w0 /path/to/local/proxy.pac)`
The code after this trick has been applied:
`// Create PAC file to send WebSocket requests direct but other protocols throught BrowserMobProxy server
String pacFunction = "function FindProxyForURL(url, host) { if (url.substring(0, 3) === "ws:" || url.substring(0, 4) === "wss:") { " +
"return "DIRECT"; } else { return "PROXY localhost:" + proxyPort + ""; } }";
seleniumProxy.setProxyAutoconfigUrl("data:application/x-javascript-config;base64," +
                        Base64.getUrlEncoder().encodeToString(pacFunction.getBytes()));`

As you can see I encoded the script not the location because that way it did not work for me.
Note: Actually I used BrowserUp proxy but I think this should work with BrowserMob also.
